### ISSUE TITLE
Enable custom "flutter create" templates

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#77104]
It would be interesting to enable the creation of custom templates usable with `flutter create`, which don't require that the templates be actually in our repo. For example, if Rive wanted to create a template that showed off how to use Rive animations in Flutter, they could do so without having to land code in the flutter/flutter repo. (VGV have done something similar: https://verygood.ventures/blog/flutter-starter-app-very-good-core-cli)

Should we enable anyone to create a template, or should we have an allowlist? There's a risk of people tricking people into running hostile code if we allow anyone to host templates. We could have two tiers, one for template sets whose creators we trust, and one which gets a big warning.

How do we identify templates? We could do it by URL, or by some well-established pattern e.g. domain:path, like `example.com:demoapp` which expands to `example.com/.flutter-templates/demoapp.zip` or some such.

How do we host templates? zip files?

How do we avoid having to include redundant files? e.g. it'd be nice if most templates could be just lib/main.dart and pubspec.yaml.

**Comment by jonahwilliams** [comment]
Some thoughts:

Many kinds of templates have tags/conditionals, but I'm not sure how either the CLI or Flutter IDEs would provide those without a lowest common denominator approach like `--template-value=foo=bar`

If most templates don't want to deal with platform configuration, then we'd need to include that by default. If we start supporting custom platform configurations, we'd need to account for them when writing migrators https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/project_migrator.dart#L15

**Comment by tarobins** [comment]
A specific template I'd like, which would be probably a way I'd use this feature, is to have a template like the app starter template but without all the comments, and maybe not even some of the code.  

When I want to just quickly try something in flutter, I end up creating and app and then deleting a bunch of the comments and some of the code cause that just clutters up my experiment work.  Not saying the comments aren't good; they are very helpful for a new users, but I'd rather `flutter create` an even more bare-bones app in some cases.

**Comment by tarobins** [comment]
A specific template I'd like, which would be probably a way I'd use this feature, is to have a template like the app starter template but without all the comments, and maybe not even some of the code.

When I want to just quickly try something in flutter, I end up creating and app and then deleting a bunch of the comments and some of the code cause that just clutters up my experiment work. Not saying the comments aren't good; they are very helpful for a new users, but I'd rather flutter create an even more bare-bones app in some cases.

[added] It would also be super helpful if I could host this template somewhere other than my local machine cause I often do flutter experiments on different machines, so I'd like to just point at some url. I realize this feature could introduce a whole load of security issues, so maybe not feasible, but anyway.  Though maybe the security issue is no different than a pub package?

**Comment by pedromassangocode** [comment]
> When I want to just quickly try something in flutter, I end up creating and app and then deleting a bunch of the comments and some of the code cause that just clutters up my experiment work.

I agree. Something like `flutter create foo --no-comments` would be great.

**Comment by Hixie** [comment]
> Though maybe the security issue is no different than a pub package?

I hadn't thought of using `pub` to host these, that's brilliant. It would reduce all the problems here to a previously solved problem. (Or at least, a previously-existing problem.)

@jonahwilliams, regarding tags/conditionals, can you elaborate? What tags and conditionals do we use in the current templates? I suspect it might be ok to not support this generally for the new templates but I am not sure if I'm missing something obvious.

**Comment by jonahwilliams** [comment]
For example, suppose your template generates a test file alongside your project. Its generally the case that we use package imports for test files, so the import would need to configured based on the chosen project name.

**Comment by Hixie** [comment]
I guess the question is whether we need such flags beyond a standard set like "project name".

**Comment by AliHG** [comment]
Are these templates gonna contain guide lines for app architecture and state management like in Jetpack MVVM ??

**Comment by AliHG** [comment]
@Hixie I saw your talk in flutter engage but i don't know if this is what you meant or not

**Comment by Hixie** [comment]
They could, but the one I mentioned during the Q&A is actually in review right now and wouldn't use this (it would just be built in). That should land soon and should be in the next stable.

**Comment by AliHG** [comment]
could you provide me a link please , thanks in advance

**Comment by AliHG** [comment]
@Hixie could you provide me the link for it

**Comment by Hixie** [comment]
Yeah, using pub.dev seems like the best proposal so far. We could literally just have these be a special kind of package: you run `flutter create --from-package foobar` and it fetches and expands the `foobar` package from pub. Might make sense for these to be an overlay on top of the existing templates so that people can just include the parts they want.

**Comment by RandalSchwartz** [comment]
> When I want to just quickly try something in flutter, I end up creating and app and then deleting a bunch of the comments and some of the code cause that just clutters up my experiment work. 

I imagine more lines have been collectively deleted from the counter app than have gone into any other single codebase. :)

**Comment by AlmogRnD** [comment]
Following, I would also like to see this. My goal would be to have a way to set up a template project with files, folders, structure, packages maybe even initial routing that can be generated via a command line like Yoman in JS. Reduce the boilerplate of setup, currently working on my 6th app in Flutter.

**Comment by felangel** [comment]
The functionality you're describing is very similar to what we have done in [very_good_cli](https://pub.dev/packages/very_good_cli) and regarding the template creation and generation we use [package:mason](https://pub.dev/packages/mason). Hope that helps! 👍

**Comment by AlmogRnD** [comment]
@felangel thanks

**Comment by AlmogRnD** [comment]
@felangel any tutorials on how you did this

**Comment by felangel** [comment]
> @felangel any tutorials on how you did this

I'd recommend taking a look at the [mason README](https://github.com/felangel/mason/blob/master/README.md) and [example](https://github.com/felangel/mason/tree/master/example). We're working on a more detailed tutorial which should be available in the next few weeks but in the meantime I'm happy to answer any questions.

**Comment by YazeedAlKhalaf** [comment]
Publishing templates to `pub.dev` should be limited to accounts that are not new, that way people don't spam the `pub.dev` with weird templates. Also, personal templates might be a good idea to use GitHub to host the files?

some thoughts 💭

**Comment by felangel** [comment]
FYI we just published a [blog post on mason](https://verygood.ventures/blog/code-generation-with-mason) 😄

**Comment by Hixie** [comment]
If you are interested in implementing this please feel free to jump in! For something like this it's probably a good idea to write a design doc first; see https://github.com/flutter/flutter/wiki/Chat#design-documents.

**Comment by jonahwilliams** [comment]
And to @Hixie 's point on a design doc - this isn't really a formality. I am personally not convinced that custom create templates in the flutter tool add a lot of value over a new tool like the very_good_cli. You should be prepared to make the case that isn't true, and to additionally explain how it would be maintained without breaking as the tool libraries/source update over time.

**Comment by lyleLH** [comment]
Just like the Cocoapods does

```
 pod lib create --template-url=<git repo url>
```
So I can use this command to create a semi-finished product App ,and that contains many 

useful dependencies and architecture code .

It's great.

**Comment by felangel** [comment]
Just wanted to give an update...

I've been actively working on [package:mason](https://pub.dev/packages/mason) and [package:mason_cli](https://pub.dev/packages/mason_cli) over the last year to make it easy for developers to define/consume reusable templates. With `mason`, developers can create and distribute templates (bricks) either via local path, git, or remote registry (heavily inspired by pub.dev, stagehand, cookie cutter, and more.). Templates can be bundled and baked into existing CLIs as well. [very_good_cli](https://pub.dev/packages/very_good_cli) is an example of a Dart CLI that leverages `mason` internally for the templating/rendering of bundled templates.

I've would love to collaborate with the Flutter team to leverage an open source tool like `mason` within `package:flutter_tools` because `mason` already offers support for many of the requested features that would make using/defining custom flutter templates easier.

**Comment by omishah** [comment]
this is really an important feature that should be available 👍

**Comment by Likhachev17** [comment]
Hi! Any updates? It's so important!

**Comment by Tananga** [comment]
This is a great idea! If I can build app based on skeleton template I also should be able to create my own custom template

**Comment by spydon** [comment]
I agree that this would be a great idea! Our use-case would be to provide a simple Flame template with a skeleton for our Flame Game Jam (or just in general), instead of having to use third-party tools like Mason or Ignite.
